### DISCUSSION TITLE
Support MQTT QoS 2 package flow

### DISCUSSION THREAD

**Discussion by WouterHouwen** [rabbitmq/rabbitmq-server#10376] [Other] [answered]
### Is your feature request related to a problem? Please describe.

We want to switch from the Mosquitto MQTT broker to RabbitMQ but have a lot of clients in the field that publish MQTT messages with QoS 2. When the client publishes a message with Qos 2, their connection gets closed, presumably because the MQTT broker returns a PUBACK package instead of a PUBREC. 

Our clients uses the C++ library libmosquittopp.
We don't have to option to update the clients in the field our selves.

### Describe the solution you'd like

When enabeling a setting, have RabbitMQ MQTT plugin, follow the QoS 2 package flow(PUBLISH -> PUBREC -> PUBREL -> PUBCOMP) when a message is published with QoS 2.

Enabling this setting will only enable the QoS 2 package flow without the "exactly once" promise by RabbitMQ.

### Describe alternatives you've considered

We have no option to solve this on the client side as we don't have an option to update the clients in the field our selves.

### Additional context

Publishing QoS 2 messages with `mosquitto_pub` also throws a `network protocol error`:
```
mosquitto_pub -h localhost -p 1883 -u dev -P dev  -t "/test" -q 2 -m "test 3" -d
Client test sending CONNECT
Client test received CONNACK (0)
Client test sending PUBLISH (d0, q2, r0, m1, '/test', ... (6 bytes))
Client test received PUBACK (Mid: 1, RC:0)
Error: A network protocol error occurred when communicating with the broker.
```

QoS package flow:  
![image](https://github.com/rabbitmq/rabbitmq-server/assets/42268945/4fee9046-9eef-4713-801f-2c6587892de3)

**Comment by michaelklishin** [comment, marked as answer]
I'm afraid that's not how things work: if we claim to support the protocol part of QoS 2, we must provide at least a **very significant majority of the guarantees** behind it, and we don't plan to, at least at the moment. 

To not recipe the entire discussion one more time, our team generally does not believe that QoS 2 provides any additional data safety in practice. So we spend our time on other things, like supporting MQTTv5 (shipping in RabbitMQ 3.13), reworking AMQP 1.0 from scratch and moving it into the core, and so on.

We fully admit that this means that some folks, for valid or "check box ticking" reasons, will have to look elsewhere.

**Reply by objectboy2016** [reply]
We believe that QoS 2 does not provide stronger security guarantees — it only ensures message delivery exactly once. Since QoS 1 can often result in duplicate message receptions, using QoS 1 requires additional logic on the client side to handle duplicates. However, many remote devices cannot implement such logic, so QoS 2 is indeed necessary in those cases.
